Fleet search wait for checkpoint timeout results in a 500 status code.

Fleet search requests fail with 500 status code if `wait_for_checkpoints` is specified and it times out. A non 5xx status code is expected if waiting for checkpoints times out.

Stacktrace:

```
Failed to execute phase [query], all shards failed; shardFailures {[fYmmBR4MQLq61HieTPfX0g][.fleet-policies-7][0]: org.elasticsearch.transport.RemoteTransportException: [es-es-search-7d99cb7764-wqf5p][100.64.207.91:9300][indices:data/read/search[phase/query]]
Caused by: org.elasticsearch.ElasticsearchTimeoutException: Wait for seq_no [2] refreshed timed out [30s]
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.lambda$ensureAfterSeqNoRefreshed$4(SearchService.java:595)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:925)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.threadpool.ThreadPool$1.run(ThreadPool.java:514)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
}
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:727)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:417)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:759)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:510)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:338)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:64)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:75)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:54)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:719)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1775)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1508)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1642)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1617)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:44)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:45)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.lambda$ensureAfterSeqNoRefreshed$4(SearchService.java:594)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:925)
	at org.elasticsearch.server@9.0.0/org.elasticsearch.threadpool.ThreadPool$1.run(ThreadPool.java:514)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: org.elasticsearch.ElasticsearchTimeoutException: Wait for seq_no [2] refreshed timed out [30s]
	at org.elasticsearch.server@9.0.0/org.elasticsearch.search.SearchService.lambda$ensureAfterSeqNoRefreshed$4(SearchService.java:595)
	... 8 more
```

--- a/server/src/main/java/org/elasticsearch/search/SearchService.java
+++ b/server/src/main/java/org/elasticsearch/search/SearchService.java
@@ -17,7 +17,6 @@ import org.apache.lucene.search.MatchAllDocsQuery;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.search.TopDocs;
 import org.elasticsearch.ElasticsearchException;
-import org.elasticsearch.ElasticsearchTimeoutException;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.ActionRunnable;
 import org.elasticsearch.action.ResolvedIndices;
@@ -33,6 +32,7 @@ import org.elasticsearch.common.CheckedSupplier;
 import org.elasticsearch.common.UUIDs;
 import org.elasticsearch.common.breaker.CircuitBreaker;
 import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.logging.LoggerMessageFormat;
 import org.elasticsearch.common.lucene.Lucene;
 import org.elasticsearch.common.settings.Setting;
 import org.elasticsearch.common.settings.Setting.Property;
@@ -112,6 +112,7 @@ import org.elasticsearch.search.query.QueryPhase;
 import org.elasticsearch.search.query.QuerySearchRequest;
 import org.elasticsearch.search.query.QuerySearchResult;
 import org.elasticsearch.search.query.ScrollQuerySearchResult;
+import org.elasticsearch.search.query.SearchTimeoutException;
 import org.elasticsearch.search.rank.feature.RankFeatureResult;
 import org.elasticsearch.search.rank.feature.RankFeatureShardPhase;
 import org.elasticsearch.search.rank.feature.RankFeatureShardRequest;
@@ -598,9 +599,13 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
             final TimeValue timeout = request.getWaitForCheckpointsTimeout();
             final Scheduler.ScheduledCancellable timeoutTask = NO_TIMEOUT.equals(timeout) ? null : threadPool.schedule(() -> {
                 if (isDone.compareAndSet(false, true)) {
-                    listener.onFailure(
-                        new ElasticsearchTimeoutException("Wait for seq_no [{}] refreshed timed out [{}]", waitForCheckpoint, timeout)
+                    var shardTarget = new SearchShardTarget(
+                        shard.routingEntry().currentNodeId(),
+                        shard.shardId(),
+                        request.getClusterAlias()
                     );
+                    var message = LoggerMessageFormat.format("Wait for seq_no [{}] refreshed timed out [{}]", waitForCheckpoint, timeout);
+                    listener.onFailure(new SearchTimeoutException(shardTarget, message));
                 }
             }, timeout, EsExecutors.DIRECT_EXECUTOR_SERVICE);
 

